### DISCUSSION TITLE
Hardware timer inclusion?

### DISCUSSION THREAD

**Discussion by Lafred** [contrem/arduino-timer#30] [Ideas]
Hi,

I was working with MillisTimer and found it only ran after a timer.run() and would not repeat.  However it is no longer actively supported.  I'm an electrical engineer and am working towards making an arduino multi timer for personal time management, to better learn the community, and to have fun with Arduino.  

What are your thoughts on making this work with the chip hardware timers?  I'm curious if there is a way to better implement the MillisTimer library and integrate the hardware timer with your class arduino-timer.  

The current arduino-timer library is probably more than enough for the functionality I need at present but I think it would be fun to learn to use the arduino hardware timers.

I'd like to collaborate on this.

James Stenquist

**Comment by contrem** [comment]
Hi, James,

Thanks for your interest in the arduino-timer library.

The reason it's currently a software only timer is because hardware timers are hardware specific. Each chip can need different code to setup and use the different interrupts for their timers. Maintaining compatibility with all the different chips is a big task, and not one I can support both time-wise and hardware-wise.

That being said, it is an interesting topic, and you should learn about using hardware timers for when you do need them. I believe there are also existing libraries that try and provide cross-platform timer APIs if you're interested in seeing example implementations.

**Reply by Lafred** [reply]
Thanks contrem,

Time is important and devices need a registers set in order to get the hardware timer working.  It takes a lot of research to get hardware timers working correctly I would think.  

If you'd like, you could take a look at my repo here where I switched to using your class for the timing elements.  
https://bitbucket.org/Lafred/multitimer/src/master/

The functionality is really beginning to come through.  The display counts down and the buzzer chirps at the end.  

My current problem is my custom CountDownTime class is not reassigning a static variable to change the timer side.  No pressure though if you'd like to take a look that would be appreciated.  

This project is for fun and to help me better manage time.  I plan to 3D print it, once the programming is finished and the model is created.

**Reply by Lafred** [reply]
Found out the problem.  What does this line do in c++?  

0 < currentAngle <= angleOfRotation

Anyway, changed it to:
0 < currentAngle && currentAngle <= angleOfRotation

**Reply by contrem** [reply]
> What does this line do in c++?

According to [c++ operator precedence](https://en.cppreference.com/w/cpp/language/operator_precedence), it would evaluate ```0 < currentAngle``` to true or false, and then evaluate ```result_of_previous <= angleOfRotation```, which is not what you want.

Side note, make sure your timer handlers return true if they need to keep running, and false otherwise. Your ```bool writeToDisplay()``` function doesn't return true.
